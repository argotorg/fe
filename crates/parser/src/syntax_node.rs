use crate::SyntaxKind;

#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub enum FeLang {}

impl rowan::Language for FeLang {
    type Kind = SyntaxKind;

    fn kind_from_raw(raw: rowan::SyntaxKind) -> Self::Kind {
        SyntaxKind::try_from(raw.0).expect("invalid SyntaxKind from raw u16")
    }

    fn kind_to_raw(kind: Self::Kind) -> rowan::SyntaxKind {
        kind.into()
    }
}

pub type SyntaxNode = rowan::SyntaxNode<FeLang>;
pub type SyntaxToken = rowan::SyntaxToken<FeLang>;
pub type GreenNode = rowan::GreenNode;
pub type TextRange = rowan::TextRange;
pub type NodeOrToken = rowan::NodeOrToken<SyntaxNode, SyntaxToken>;
