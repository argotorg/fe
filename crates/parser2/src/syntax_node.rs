use crate::SyntaxKind;

#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub enum FeLang {}

impl rowan::Language for FeLang {
    type Kind = SyntaxKind;

    fn kind_from_raw(raw: rowan::SyntaxKind) -> Self::Kind {
        unsafe { std::mem::transmute::<u16, SyntaxKind>(raw.0) }
    }

    fn kind_to_raw(kind: Self::Kind) -> rowan::SyntaxKind {
        kind.into()
    }
}

pub type SyntaxNode = rowan::SyntaxNode<FeLang>;
