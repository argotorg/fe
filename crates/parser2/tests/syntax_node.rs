use fe_parser2::syntax_node::SyntaxNode;

#[allow(unused)]
fn build_cst(input: &str) -> SyntaxNode {
    let (cst, errors) = fe_parser2::parse_source_file(input);
    for error in &errors {
        println!("{}", error.msg);
    }
    assert! {errors.is_empty()}
    cst
}

fe_compiler_test_utils::build_debug_snap_tests! {
    "parser2/test_files/syntax_node",
    "parser2/test_files/syntax_node",
    build_cst
}
