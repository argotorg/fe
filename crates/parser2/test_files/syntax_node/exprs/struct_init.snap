---
source: crates/parser2/tests/syntax_node.rs
expression: snapshot
---
Root@0..31
  RecordInitExpr@0..13
    Path@0..6
      PathSegment@0..6
        Ident@0..6 "Struct"
    WhiteSpace@6..7 " "
    RecordFieldList@7..13
      LBrace@7..8 "{"
      RecordField@8..9
        Ident@8..9 "x"
      Comma@9..10 ","
      WhiteSpace@10..11 " "
      RecordField@11..12
        Ident@11..12 "y"
      RBrace@12..13 "}"
  Newline@13..14 "\n"
  RecordInitExpr@14..31
    Path@14..20
      PathSegment@14..20
        Ident@14..20 "Struct"
    WhiteSpace@20..21 " "
    RecordFieldList@21..31
      LBrace@21..22 "{"
      RecordField@22..30
        Ident@22..23 "x"
        Colon@23..24 ":"
        WhiteSpace@24..25 " "
        BinExpr@25..30
          LitExpr@25..26
            Int@25..26 "1"
          WhiteSpace@26..27 " "
          Plus@27..28 "+"
          WhiteSpace@28..29 " "
          LitExpr@29..30
            Int@29..30 "2"
      RBrace@30..31 "}"

